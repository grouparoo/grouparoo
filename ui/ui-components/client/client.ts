import Axios, {
  AxiosRequestConfig,
  AxiosRequestHeaders,
  AxiosResponse,
  Method,
} from "axios";
import type { IncomingMessage, ServerResponse } from "http";

import { isBrowser } from "../utils/isBrowser";
import PackageJSON from "../package.json";
import { errorHandler } from "../eventHandlers";
import type { AppContext } from "next/app";
import type { GetServerSidePropsContext, NextPageContext } from "next";
import { getRequestContext } from "../utils/appContext";
import type { ErrorHandler } from "../eventHandlers/errorHandler";

interface ClientCacheObject {
  locked: boolean;
  data: any;
  createdAt: number;
}
export class ClientCache {
  cache: { [key: string]: ClientCacheObject };
  ttl: number;

  constructor(ttl = 5000) {
    this.ttl = ttl;
    this.cache = {};
  }

  async sleep(wait = 100) {
    return new Promise((resolve) => {
      setTimeout(resolve, wait);
    });
  }

  async get(key: string) {
    if (!this.cache[key]) {
      this.cache[key] = { locked: false, data: undefined, createdAt: 0 };
    }

    let waitCounter = 0;
    while (this.cache[key]?.locked && waitCounter < 15) {
      await this.sleep(300);
      waitCounter++;
    }

    const now = new Date().getTime();
    if (this.cache[key]?.data) {
      if (this.cache[key].createdAt + this.ttl < now) {
        delete this.cache[key];
        this.cache[key] = { locked: true, data: undefined, createdAt: 0 };
      }
    } else {
      this.cache[key] = { locked: true, data: undefined, createdAt: 0 };
    }

    const unlock = (data) => {
      this.cache[key] = {
        data,
        createdAt: new Date().getTime(),
        locked: false,
      };
    };

    return { cacheData: this.cache[key]?.data, unlock };
  }

  clear() {
    this.cache = {};
  }
}

interface ClientRequestOptions {
  useCache: boolean;
  errorHandler: ErrorHandler;
}

export class Client {
  public apiVersion = process.env.API_VERSION || "v1";

  private webUrl = process.env.WEB_URL || "";
  private serverToken = process.env.SERVER_TOKEN;
  private cache = new ClientCache();

  private static readonly optionDefaults: ClientRequestOptions = {
    useCache: true,
    errorHandler,
  };

  constructor(
    private getRequestContext: () => {
      req?: IncomingMessage;
      res?: ServerResponse;
    } = () => ({})
  ) {}

  private csrfToken = () => {
    if (globalThis?.localStorage) {
      return window.localStorage.getItem("session:csrfToken");
    }
  };

  public request = async <Response = any>(
    verb: Method = "get",
    path: string,
    data: AxiosRequestConfig["data"] = {},
    options: Partial<ClientRequestOptions> = {}
  ): Promise<Response> => {
    options = { ...Client.optionDefaults, ...options };

    const headers: AxiosRequestHeaders = {
      Accept: "application/json",
      "Content-Type": "application/json",
      "X-Grouparoo-Client": `${PackageJSON.name}-v${PackageJSON.version}`,
    };

    const { req } = this.getRequestContext();
    if (req?.headers?.cookie) {
      headers["X-GROUPAROO-SERVER-TOKEN"] = this.serverToken;
      headers["cookie"] = req?.headers?.cookie;
      options.useCache = false; // do not ever cache responses on the server
    }

    const config: AxiosRequestConfig = {
      params: null,
      data: null,
      url: `${this.webUrl}/api/${this.apiVersion}${path}`, // path comes with a leading "/"
      withCredentials: true,
      method: verb.toLowerCase() as Method,
      headers,
    };

    data.csrfToken = this.csrfToken();

    if (data) {
      for (const i in data) {
        if (data[i] === null || data[i] === undefined) {
          delete data[i];
        }
      }

      if (data.file || data._file) {
        delete config.headers;
        let dataForm = new FormData();
        for (const i in data) {
          dataForm.append(i, data[i]);
        }
        data = dataForm;
      }

      if (config.method === "get") {
        config.params = data;
      } else {
        config.data = data;
      }
    }

    let unlock: Function;
    if (config.method === "get" && options.useCache) {
      const { cacheData, unlock: _unlock } = await this.cache.get(
        config.url + JSON.stringify(data)
      );
      unlock = _unlock;
      if (cacheData) {
        return cacheData;
      }
    } else {
      this.cache.clear();
    }

    try {
      const response: AxiosResponse<Response & { error?: Error }> = await Axios(
        config
      );

      if (unlock) {
        unlock(response.data);
      }

      if (response.data && response.data.error) {
        throw response.data.error.message
          ? response.data.error.message
          : response.data.error;
      }
      return response.data;
    } catch (error) {
      if (unlock) {
        unlock();
      }

      if (error.response && error.response.data && error.response.data.error) {
        const err =
          error.response?.data?.error?.message ??
          error.response?.data?.error ??
          error;

        options.errorHandler.set({ message: err });
        const newError = Error(err);
        newError["code"] = error.response?.data?.error?.code;
        throw newError;
      } else {
        options.errorHandler.set({ message: error });
        throw error;
      }
    }
  };
}

export const generateClient = (
  ctx: AppContext | NextPageContext | GetServerSidePropsContext
) => new Client(getRequestContext(ctx));
