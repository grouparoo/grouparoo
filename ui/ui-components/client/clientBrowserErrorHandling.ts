import axios from "axios";
import { errorHandler } from "../eventHandlers";
import { getRedirectFromErrorCode } from "../utils/getRedirectFromErrorCode";
import { isBrowser } from "../utils/isBrowser";

if (isBrowser()) {
  axios.interceptors.response.use(undefined, (error) => {
    errorHandler.set({ message: error });

    const redirect = getRedirectFromErrorCode(
      error.code,
      window.location.pathname
    );

    if (redirect) {
      window.location.href = redirect.destination;
      return new Promise(() => {});
    }

    return {};
  });
}
