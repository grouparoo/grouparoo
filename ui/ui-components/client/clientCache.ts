export interface ClientCacheObject<T = unknown> {
  locked: boolean;
  data: T;
  createdAt: number;
}

export interface ClientCacheGetObject<T = unknown> {
  cacheData: T;
  unlock: (data?: T) => void;
}

export class ClientCache {
  private cache: Record<string, ClientCacheObject> = {};

  constructor(private ttl: number = 5000) {}

  async sleep(wait = 100) {
    return new Promise((resolve) => {
      setTimeout(resolve, wait);
    });
  }

  async get<T = unknown>(key: string): Promise<ClientCacheGetObject<T>> {
    if (!this.cache[key]) {
      this.cache[key] = { locked: false, data: undefined, createdAt: 0 };
    }

    let waitCounter = 0;
    while (this.cache[key]?.locked && waitCounter < 15) {
      await this.sleep(300);
      waitCounter++;
    }

    const now = new Date().getTime();
    if (this.cache[key]?.data) {
      if (this.cache[key].createdAt + this.ttl < now) {
        delete this.cache[key];
        this.cache[key] = { locked: true, data: undefined, createdAt: 0 };
      }
    } else {
      this.cache[key] = { locked: true, data: undefined, createdAt: 0 };
    }

    const unlock = (data?: T) => {
      this.cache[key] = {
        data,
        createdAt: new Date().getTime(),
        locked: false,
      };
    };

    return { cacheData: this.cache[key]?.data as T, unlock };
  }

  clear() {
    this.cache = {};
  }
}
