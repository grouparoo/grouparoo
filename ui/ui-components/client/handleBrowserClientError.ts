import { errorHandler } from "../eventHandlers";
import { getRedirectFromErrorCode } from "../utils/getRedirectFromErrorCode";

export const handleBrowserClientError = async <Response>(
  error: Error & { code?: string }
): Promise<Response> => {
  errorHandler.set({ message: error });

  const redirect = getRedirectFromErrorCode(
    error.code,
    window.location.pathname
  );

  if (redirect) {
    window.location.href = redirect.destination;
    return new Promise(() => {});
  }

  return {} as Response;
};
