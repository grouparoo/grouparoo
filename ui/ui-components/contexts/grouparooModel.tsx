import { createContext, useContext } from "react";
import { Models } from "../utils/apiData";

interface GrouparooModelContext {
  model: Models.GrouparooModelType;
}

export const GrouparooModelContext =
  createContext<GrouparooModelContext>(undefined);

export const useGrouparooModel = () => useContext(GrouparooModelContext);
