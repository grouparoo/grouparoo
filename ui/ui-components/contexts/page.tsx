import { createContext, useContext } from "react";
import { Actions } from "../utils/apiData";

export interface PageContext {
  clusterName: Actions.NavigationList["clusterName"];
  currentTeamMember: Partial<Actions.NavigationList["teamMember"]>;
  navigation: Actions.NavigationList["navigation"];
  navigationMode: Actions.NavigationList["navigationMode"];
}

export const PageContext = createContext<PageContext>(undefined);

export const usePageContext = () => useContext(PageContext);
