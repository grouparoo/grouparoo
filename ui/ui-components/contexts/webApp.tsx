import { createContext, useContext } from "react";
import { Actions } from "../utils/apiData";

export interface WebAppContext {
  clusterName: Actions.NavigationList["clusterName"];
  currentTeamMember: Partial<Actions.NavigationList["teamMember"]>;
  navigation: Actions.NavigationList["navigation"];
  navigationMode: Actions.NavigationList["navigationMode"];
}

export const WebAppContext = createContext<WebAppContext>(undefined);

export const useWebAppContext = () => useContext(WebAppContext);
