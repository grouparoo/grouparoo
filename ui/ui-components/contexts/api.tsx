import { createContext, useContext } from "react";
import { Client } from "../client/client";

export const ApiContext = createContext({
  client: new Client(),
});
export const useApi = () => useContext(ApiContext);
