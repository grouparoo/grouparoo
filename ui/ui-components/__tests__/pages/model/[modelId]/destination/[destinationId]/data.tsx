import { Client } from "../../../../../../client/client";
import { ApiContext } from "../../../../../../contexts/api";
import { GrouparooModelContext } from "../../../../../../contexts/grouparooModel";
import DataPage, {
  getServerSideProps,
} from "../../../../../../pages/model/[modelId]/destination/[destinationId]/data";
import { renderAndWait } from "../../../../../__utils__/renderAndWait";

const destinationId = "test-destination";
const modelId = "test-model";
const groupId = "test-group";

jest.mock("next/router", () => {
  return {
    useRouter: () => ({ query: { destinationId }, asPath: "", pathname: "" }),
  };
});

jest.mock(
  "../../../../../../components/destination/DestinationSampleRecord",
  () => () => null
);

jest.mock("../../../../../../client/client", () => {
  const fakeRequest = () =>
    Promise.resolve({
      destination: {
        destinationId,
        modelId,
        group: {
          groupId,
        },
        mapping: {},
        app: {
          icon: "",
        },
      },
      groups: [],
      properties: [],
      options: {
        properties: {
          known: [],
          required: [],
        },
      },
    });
  class FakeClient {
    constructor() {
      return {
        request: fakeRequest,
      };
    }
  }

  return {
    Client: FakeClient,
    generateClient: () => new FakeClient(),
  };
});

describe("model destination data page", () => {
  it("should render", async () => {
    const props: any = await getServerSideProps({
      query: {
        destinationId,
        modelId,
      },
    } as any);

    const dataPage = await renderAndWait(
      <GrouparooModelContext.Provider value={{ model: {} }}>
        <ApiContext.Provider value={{ client: new Client() }}>
          <DataPage {...props.props} />
        </ApiContext.Provider>
      </GrouparooModelContext.Provider>
    );
    expect(dataPage.container).toMatchSnapshot();
  });
});
