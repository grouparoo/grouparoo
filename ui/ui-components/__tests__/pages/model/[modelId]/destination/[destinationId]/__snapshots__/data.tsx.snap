// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model destination data page should render 1`] = `
<div>
  <div
    class="mb-3 media"
  >
    <div
      style="padding: 10px; text-align: center; height: 100px; width: 100px;"
    >
      <br />
      <span
        class="badge"
      >
        <em>
          no icon
        </em>
      </span>
    </div>
    <div
      class="media-body"
    >
      <div>
        <h1
          class="mr-3"
          style="display: inline;"
        >
          <em>
            Draft
          </em>
        </h1>
      </div>
        
      <span
        class="badge badge-primary"
        style="margin-left: 3px; margin-right: 3px; margin-bottom: 20px;"
      />
        
      <span
        class="badge badge-primary"
        style="margin-bottom: 20px;"
      >
        <span>
          <a
            style="color: white;"
          />
        </span>
      </span>
      
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="mb-4 col"
    >
      <form
        class=""
        id="form"
      >
        <fieldset>
          <div
            class="row"
            id="destinationCollection"
          >
            <div
              class="col"
            >
              <h5>
                Who should be sent to
                 
                <span
                  class="text-primary"
                />
                ?
              </h5>
              <select
                class="form-control"
                required=""
              >
                <option
                  value="__none"
                >
                  No Model or Group
                </option>
                <option
                  disabled=""
                >
                  --- Models ---
                </option>
                <option
                  value="__model"
                >
                  All Records in the  Model
                </option>
                <option
                  disabled=""
                >
                  --- Groups ---
                </option>
              </select>
            </div>
          </div>
          <br />
          <div
            class="row"
            id="destinationData"
          >
            <div
              class="col"
            >
              <h5>
                What data do you want in
                 
                <span
                  class="text-primary"
                />
                ?
              </h5>
              <br />
              <div
                data-screenshotid="destinationRecords"
              />
              <br />
              <div
                data-screenshotid="destinationGroups"
              />
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <br />
              <hr />
              <button
                class="btn btn-primary"
                style="display: inline-flex; align-items: center; justify-content: center;"
                type="submit"
              >
                <span
                  style="opacity: 1;"
                >
                  Save Destination Data
                </span>
              </button>
            </div>
          </div>
        </fieldset>
      </form>
    </div>
    <div
      class="col-xl-5"
    />
  </div>
</div>
`;
